drivers/mtd/mtd_config_fs.c

```diff
@@ -0,0 +1,2484 @@
+
+  fs->data_wra = fs->ate_wra & ADDR_SECT_MASK;
+  finfo("sector close, data_wra=0x%lx\n", fs->data_wra);
+
+  return 0;
+}
+
+/****************************************************************************
+ * Name: nvs_add_gc_done_ate
+ ****************************************************************************/
+
+static int nvs_add_gc_done_ate(FAR struct nvs_fs *fs)
+{
+  struct nvs_ate gc_done_ate;
+
+  finfo("Adding gc done ate at %lx\n", fs->ate_wra & ADDR_OFFS_MASK);
+  gc_done_ate.id = NVS_SPECIAL_ATE_ID;
+  gc_done_ate.len = 0U;
+  gc_done_ate.key_len = 0U;
+  gc_done_ate.offset = (uint16_t)(fs->data_wra & ADDR_OFFS_MASK);
+  nvs_ate_crc8_update(&gc_done_ate);
+
+  return nvs_flash_ate_wrt(fs, &gc_done_ate);
+}
+
+/****************************************************************************
+ * Name: nvs_expire_ate
+ ****************************************************************************/
+
+static int nvs_expire_ate(FAR struct nvs_fs *fs, uint32_t addr)
+{
+  uint8_t expired = 0x7f;
+  off_t offset;
+
+  offset = fs->geo.erasesize * (addr >> ADDR_SECT_SHIFT);
+  offset += addr & ADDR_OFFS_MASK;
+
+  return flash_write(fs->mtd, offset + offsetof(struct nvs_ate, expired),
+                     &expired, sizeof(expired));
+}
+
+/****************************************************************************
+ * Name: nvs_gc
+ *
+ * Description:
+ *   garbage collection: the address ate_wra has been updated to the new
+ *   sector that has just been started. The data to gc is in the sector
+ *   after this new sector.
+ *
+ ****************************************************************************/
+
+static int nvs_gc(FAR struct nvs_fs *fs)
+{
+  int rc;
+  struct nvs_ate close_ate;
+  struct nvs_ate gc_ate;
+  uint32_t sec_addr;
+  uint32_t gc_addr;
+  uint32_t gc_prev_addr;
+  uint32_t data_addr;
+  uint32_t stop_addr;
+
+  finfo("gc: before gc, ate_wra %lx, ate_size %d\n", fs->ate_wra,
+    sizeof(struct nvs_ate));
+
+  sec_addr = (fs->ate_wra & ADDR_SECT_MASK);
+  nvs_sector_advance(fs, &sec_addr);
+  gc_addr = sec_addr + fs->geo.erasesize - sizeof(struct nvs_ate);
+
+  finfo("gc: set, sec_addr %lx, gc_addr %lx\n", sec_addr, gc_addr);
+
+  /* if the sector is not closed don't do gc */
+
+  rc = nvs_flash_ate_rd(fs, gc_addr, &close_ate);
+  if (rc < 0)
+    {
+      /* flash error */
+
+      return rc;
+    }
+
+  rc = nvs_ate_cmp_const(&close_ate, CONFIG_MTD_CONFIG_ERASEDVALUE);
+  if (!rc)
+    {
+      goto gc_done;
+    }
+
+  stop_addr = gc_addr - sizeof(struct nvs_ate);
+
+  if (nvs_close_ate_valid(fs, &close_ate))
+    {
+      gc_addr &= ADDR_SECT_MASK;
+      gc_addr += close_ate.offset;
+    }
+  else
+    {
+      rc = nvs_recover_last_ate(fs, &gc_addr);
+      if (rc)
+        {
+          return rc;
+        }
+    }
+
+  do
+    {
+      gc_prev_addr = gc_addr;
+      rc = nvs_prev_ate(fs, &gc_addr, &gc_ate);
+      if (rc)
+        {
+          return rc;
+        }
+
+      if (gc_ate.expired != 0xff)
```

Comment:
not always 0xff

Reply:
Yes, it is not always 0xff. Any value other than 0xff means the ate is an expired one.
This is to prevent a power loss when we are expiring an ate, at such case, only parts of bits are written. But we can tell it is  expired if it is not 0xff.
Perhaps CONFIG_MTD_CONFIG_ERASEDVALUE is better. that is my intention.